### PR TITLE
gh-115480: Type propagate _BINARY_OP_ADD_UNICODE

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
* Issue: gh-115480

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
Needs https://github.com/python/cpython/pull/115686 to be merged first.

File: Python/tier2_redundancy_eliminator_bytecodes.c [hunk 1]
```diff
@@ -197,6 +197,17 @@ dummy_func(void) {
         }
     }
 
+    op(_BINARY_OP_ADD_UNICODE, (left, right -- res)) {
+        if (is_const(left) && is_const(right)) {
+            PyObject *temp = PyUnicode_Concat(get_const(left), get_const(right));
+            ERROR_IF(temp == NULL, error);
+            OUT_OF_SPACE_IF_NULL(res = sym_new_const(ctx, temp));
```
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: This worries me about memory leakage. If we have a valid concatenation result but the `sym_new_const()` ran out of space, we leak the result. Even worse, that is documented as taking a borrowed reference. I *think* that's what we normally call *stealing* a reference? *Borrowed* references are normally return values (indicating that the caller shouldn't DECREF it) but I believe for arguments we don't use borrowed.

One way out (assuming the intention is for `sym_new_const()` to take ownership of its `const_val` argument) is to make that function `DECREF(const_val)` when it returns NULL.

File: Python/tier2_redundancy_eliminator_bytecodes.c [hunk 2]
```diff
@@ -197,6 +197,17 @@ dummy_func(void) {
...
+        }
+        else {
+            OUT_OF_SPACE_IF_NULL(res = sym_new_known_type(ctx, &PyUnicode_Type));
```
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: FWIW I am still in favor of splitting all calls to `OPT_OF_SPACE_IF_NULL()` into two, one with the side effect (`res = ...` etc.) and one with the error check. Slightly more verbose but doesn't bury the lede.

File: Lib/test/test_capi/test_opt.py [hunk 3]
```diff
@@ -775,10 +775,13 @@ def testfunc(n):
...
             for _ in range(n):
                 a = a + 0.1
+                a = a + 0.1
+                a = a + 0.1
+                a = a + 0.1
             return a
 
         res, ex = self._run_with_optimizer(testfunc, 32)
-        self.assertAlmostEqual(res, 4.2)
+        self.assertAlmostEqual(res, 13.8)
```
Comment (gvanrossum) [inline Lib/test/test_capi/test_opt.py]: Could you change all the float examples to use fractional parts that are powers of 2? (E.g. 0.5, 0.25, 0.125.) Adding 0.1 + 0.1 + 0.1 + 0.1 and expecting the outcome to be 0.4 is just not safe, due to float rounding and decimal/binary issues.

File: Python/tier2_redundancy_eliminator_bytecodes.c (hunk 1 above)
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: I think we know how to do this right now (see gh-116062). Unless you'd first like to do a pure cleanup pass that introduces `ERROR_IF` (and maybe `BOTTOM_IF` as well?)?

File: Lib/test/test_capi/test_opt.py [hunk 4]
```diff
@@ -842,6 +851,24 @@ def testfunc(n):
...
         self.assertIn("_BINARY_OP_MULTIPLY_FLOAT", uops)
 
+    def test_add_unicode_propagation(self):
+        def testfunc(n):
+            a = ""
+            for _ in range(n):
+                a + a
+                a + a
+                a + a
+                a + a
```
Comment (gvanrossum) [inline Lib/test/test_capi/test_opt.py:861]: Sooner or later we'll have a peephole pass that will delete the generated code for this completely. Is there a reason you're not doing something like `a = a + a`?

**Review by gvanrossum — APPROVED** [review]
Let's go ahead and merge this.

File: Lib/test/test_capi/test_opt.py (hunk 4 above)
Comment (Fidget-Spinner) [reply Lib/test/test_capi/test_opt.py:861]: That transforms it to `BINARY_OP_INPLACE_ADD_UNICODE` sadly. Which is an unsupported opcode, so the trace ends there.

**Pull request merged**